core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -885,13 +885,25 @@ public CompletionStats completionStats(String... fields) {
     }
 
     /**
-     * Rolls the tranlog generation.
+     * checks and removes translog files that no longer need to be retained. See
+     * {@link org.elasticsearch.index.translog.TranslogDeletionPolicy} for details
+     */
+    public void trimTranslog() {
+        verifyNotClosed();
+        final Engine engine = getEngine();
+        engine.trimTranslog();
+    }
+
+    /**
+     * Rolls the tranlog generation and cleans unneeded.
      *
      * @throws IOException if any file operations on the translog throw an I/O exception
      */
-    private void rollTranslogGeneration() throws IOException {
+    private void rollTranslogAndTrimGeneration() throws IOException {
         final Engine engine = getEngine();
-        engine.getTranslog().rollGeneration();
+        final Translog translog = engine.getTranslog();
+        translog.rollGeneration();
+        translog.trimUnreferencedReaders();
```

Comment:
That's a great observation. Strictly speaking we don't need to coordinate through the write lock here. In the place you linked to I was mostly interested in the error handling and that we fail the engine if anything goes wrong (as we close the translog on error as well). I was worried for poisonous situations where we keep trying to get stuff from a closed translog.  I will work on removing the read lock in the engine and fold this functionality (trim and roll) into the engine so we can be consistent. agreed?